### PR TITLE
Feature: Added the possibility to blind data points

### PR DESCRIPTION
**Pull request by haddadanas**
if the sensitivity (S/sqrt(S + B) of MC processes) is above a threshold in a bin, the data point are set to 0.

The threshold can be given in the CLI or alternatively can be set to a default in the config via: `self.config_inst.x.default_blinding_threshold`

Also needed is to define the signal processes in the config via:
a dict entry called signals in the `cfg.x.process_groups`

### CODE REVIEW

File: columnflow/tasks/framework/plotting.py [hunk 1]
```diff
@@ -72,7 +72,14 @@ class PlotBase(ConfigTask):
         description="when True, an ipython debugger is started after the plot figure is created "
         "for interactive adjustments; default: False",
     )
-
+    blinding_threshold = luigi.FloatParameter(
+        default=-1.0,
```
Comment (riga) [inline columnflow/tasks/framework/plotting.py:76]: `law.NO_FLOAT` :)

File: columnflow/tasks/framework/plotting.py [hunk 2]
```diff
@@ -72,7 +72,14 @@ class PlotBase(ConfigTask):
...
+        significant=False,
+        description="parameter to blind datapoints in the region where the sensitivity exceeds a "
+        "certain threshold, calculated from the sensitivity of MC signal and background stacks; "
+        "via a dictionary in the `default_blinding_threshold` auxiliary in the config; "
```
Comment (riga) [inline columnflow/tasks/framework/plotting.py:80]: Can be removed.

File: columnflow/plotting/plot_util.py [hunk 3]
```diff
@@ -663,3 +664,41 @@ def get_profile_variations(h_in: hist.Hist, axis: int = 1) -> dict[str, hist.His
     h_down[...] = h_view
 
     return {"nominal": h_nom, "up": h_up, "down": h_down}
+
+
+def blind_sensitivity(hists: OrderedDict, config_inst: od.Config, threshold: float | None) -> hist.Hist:
```
Comment (riga) [inline columnflow/plotting/plot_util.py:669]: - `hists: dict[od.Process, hist.Hist]`
- `threshold: float`
- `-> dict[od.Process, hist.Hist]`

File: columnflow/plotting/plot_util.py [hunk 4]
```diff
@@ -663,3 +664,41 @@ def get_profile_variations(h_in: hist.Hist, axis: int = 1) -> dict[str, hist.His
...
+    """
+    Function that takes a histogram *h_in* and blinds the values of the profile
+    over the axis *axis* that are below a certain threshold *threshold*.
+    The function needs an entry in the process_groups key of the config auxiliary
+    that is called "signals" to know, where the signal processes are defined (regex allowed).
```
Comment (riga) [inline columnflow/plotting/plot_util.py:674]: Add comment about returned copies.

File: columnflow/plotting/plot_util.py [hunk 5]
```diff
@@ -663,3 +664,41 @@ def get_profile_variations(h_in: hist.Hist, axis: int = 1) -> dict[str, hist.His
...
+    """
+    # do nothing if there are no data histograms or threshold is not given
+    if not any([proc.is_data for proc in hists.keys()]) or threshold is None:
+        return hists
```
Comment (riga) [inline columnflow/plotting/plot_util.py:678]: Can be dropped.

File: columnflow/plotting/plot_util.py [hunk 6]
```diff
@@ -663,3 +664,41 @@ def get_profile_variations(h_in: hist.Hist, axis: int = 1) -> dict[str, hist.His
...
+
+    # build the regex for the signal processes
+    signals_regex = re.compile(
+        "|".join(config_inst.x.process_groups.get("signals", [])).replace("*", ".*"),
+    )
```
Comment (riga) [inline columnflow/plotting/plot_util.py:683]: Include subprocesses.

File: columnflow/plotting/plot_util.py [hunk 7]
```diff
@@ -663,3 +664,41 @@ def get_profile_variations(h_in: hist.Hist, axis: int = 1) -> dict[str, hist.His
...
+
+    # separate histograms into signals, backgrounds and data hists and calculate sums
+    signals = {proc: hist for proc, hist in hists.items() if signals_regex.match(proc.name)}
+    data = {proc: hist for proc, hist in hists.items() if proc.is_data}
+    backgrounds = {proc: hist for proc, hist in hists.items() if proc not in signals and proc.is_mc}
```
Comment (riga) [inline columnflow/plotting/plot_util.py:688]: Return hists unchanged in case any of the three dicts is empty.

**Review by riga — CHANGES_REQUESTED** [review]
Nice! Just a couple inline comments.

**Comment by haddadanas** [comment]
Feedback is implemented however naf is currently down, so I couldn't test it :eyes: 
Will do this tomorrow in the morning

File: analysis_templates/cms_minimal/__cf_module_name__/config/analysis___cf_short_name_lc__.py [hunk 8]
```diff
@@ -120,7 +120,10 @@
 
 # process groups for conveniently looping over certain processs
 # (used in wrapper_factory and during plotting)
-cfg.x.process_groups = {}
+cfg.x.process_groups = {
+    "signals": [],  # list of signal parent processes e.g. tt, hh etc. (required for some tasks)
```
Comment (riga) [inline analysis_templates/cms_minimal/__cf_module_name__/config/analysis___cf_short_name_lc__.py:124]:
```suggestion
    "signals": [],  # list of signal parent processes e.g. h, hh etc. (needed for some features)
```

File: columnflow/plotting/plot_functions_1d.py [hunk 9]
```diff
@@ -22,6 +22,7 @@
     apply_density_to_hists,
     get_position,
     get_profile_variations,
+    blind_sensitivity,
```
Comment (riga) [inline columnflow/plotting/plot_functions_1d.py:25]: Could you rename the function to `blind_sensitive_bins`?

File: columnflow/plotting/plot_util.py [hunk 10]
```diff
@@ -663,3 +664,45 @@ def get_profile_variations(h_in: hist.Hist, axis: int = 1) -> dict[str, hist.His
...
+    hists: dict[od.Process, hist.Hist],
+    config_inst: od.Config,
+    threshold: float,
+) -> dict[od.Process, hist.Hist]:
+    """
+    Function that takes a histogram *h_in* and blinds the values of the profile
+    over the axis *axis* that are below a certain threshold *threshold*.
+    The function needs an entry in the process_groups key of the config auxiliary
+    that is called "signals" to know, where the signal processes are defined (regex allowed).
+    The histograms are not changed inplace, but The function return a copy of the modified histograms.
```
Comment (riga) [inline columnflow/plotting/plot_util.py:679]:
```suggestion
    The histograms are not changed inplace, but copies of the modified histograms are returned.
```

File: columnflow/tasks/framework/plotting.py [hunk 11]
```diff
@@ -72,7 +72,13 @@ class PlotBase(ConfigTask):
         description="when True, an ipython debugger is started after the plot figure is created "
         "for interactive adjustments; default: False",
     )
-
+    blinding_threshold = luigi.FloatParameter(
+        default=law.NO_FLOAT,
+        significant=False,
+        description="parameter to blind datapoints in the region where the sensitivity exceeds a "
+        "certain threshold, calculated from the sensitivity of MC signal and background stacks; "
+        "defaults to the `default_blinding_threshold` aux field if not given via CLI",
```
Comment (riga) [inline columnflow/tasks/framework/plotting.py:80]:
```suggestion
        description="parameter to blind datapoints in the region where the sensitivity s/sqrt(s+b) "
        "exceeds a certain threshold; defaults to the `default_blinding_threshold` aux field of "
        "the config",
```

**Review by riga — CHANGES_REQUESTED** [review]
Some last comments before we can merge :)

**Review by riga — APPROVED** [review]

**Pull request merged**